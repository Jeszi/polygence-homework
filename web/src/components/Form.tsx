import React, { useState } from "react";
import { DateTime } from "luxon";

import { InputStyles, Button } from "../styles/InputStyles";
import { SelectStyles } from "../styles/SelectStyles";
import { FormStyles } from "../styles/ComponentStyles";

import { useSpendings } from "../hooks/useSpendings";

interface FormState {
  description: string;
  amount: any;
  currency: string;
}

export const Form: React.FC = () => {
  const { addSpending } = useSpendings();

  const [state, setState] = useState<FormState>({
    description: "",
    amount: 0,
    currency: "USD",
  });

  const [errors, setErrors] = useState<string[]>([]);

  function validateFields() {
    if (state.description === "") {
      setErrors((prevErrors) => [...prevErrors, "description"]);
    }

    if (state.amount === "" || state.amount === 0) {
      setErrors((prevErrors) => [...prevErrors, "amount"]);
    }
  }

  function isFieldsValid() {
    if (state.description === "") {
      return false;
    }

    if (state.amount === "" || state.amount === 0) {
      return false;
    }

    return true;
  }

  function handleChange(e) {
    const { name, value } = e.target;
    setErrors([]);
    setState({
      ...state,
      [name]: value,
    });
  }

  function handleSave() {
    if (isFieldsValid()) {
      addSpending({
        amount: state.amount,
        currency: state.currency,
        description: state.description,
        spent_at: DateTime.now().toString(),
      });
      setState({
        ...state,
        description: "",
        amount: 0,
      });
      setErrors([]);
    } else {
      validateFields();
    }
  }

  function getError(value: string) {
    return errors.includes(value);
  }

  return (
    <>
      <FormStyles>
        <InputStyles
          data-testid="description-input"
          type="text"
          placeholder="description"
          name="description"
          value={state.description}
          onChange={handleChange}
          isError={getError("description")}
        />
        <InputStyles
          data-testid="amount-input"
          type="number"
          placeholder="amount"
          name="amount"
          min={0}
          value={state.amount}
          onChange={handleChange}
          isError={getError("amount")}
        />
        <SelectStyles
          data-testid="currency-input"
          name="currency"
          value={state.currency}
          onChange={handleChange}
        >
          <option value="HUF">HUF</option>
          <option value="USD">USD</option>
        </SelectStyles>
        <Button data-testid="save-button" onClick={handleSave}>
          Save
        </Button>
      </FormStyles>
    </>
  );
};
