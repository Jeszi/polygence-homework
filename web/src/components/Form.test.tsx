import React from "react";
import { screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../testing/test-utils";

import { Form } from "./Form";

describe("Form component", () => {
  it("should render component", async () => {
    const { asFragment } = renderWithProviders(<Form />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should have errors if validation is failed", async () => {
    renderWithProviders(<Form />);
    userEvent.click(screen.getByTestId("save-button"));

    expect(screen.getByTestId("description-input")).toHaveStyle(
      "border: solid 1px red"
    );
    expect(screen.getByTestId("amount-input")).toHaveStyle(
      "border: solid 1px red"
    );
  });

  it("should not have errors if validation is passed", async () => {
    renderWithProviders(<Form />);
    fireEvent.change(screen.getByTestId("description-input"), {
      target: { value: "test" },
    });

    fireEvent.change(screen.getByTestId("amount-input"), {
      target: { value: 10 },
    });

    userEvent.click(screen.getByTestId("save-button"));

    expect(screen.getByTestId("description-input")).not.toHaveStyle(
      "border: solid 1px red"
    );
    expect(screen.getByTestId("amount-input")).not.toHaveStyle(
      "border: solid 1px red"
    );
  });

  it("should remove errors when input is focused", async () => {
    renderWithProviders(<Form />);
    userEvent.click(screen.getByTestId("save-button"));

    expect(screen.getByTestId("description-input")).toHaveStyle(
      "border: solid 1px red"
    );
    expect(screen.getByTestId("amount-input")).toHaveStyle(
      "border: solid 1px red"
    );

    fireEvent.change(screen.getByTestId("description-input"), {
      target: { value: "test" },
    });

    expect(screen.getByTestId("description-input")).not.toHaveStyle(
      "border: solid 1px red"
    );
  });
});
