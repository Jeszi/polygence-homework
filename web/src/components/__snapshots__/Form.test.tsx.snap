// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render component 1`] = `
<DocumentFragment>
  <form
    class="sc-furwcr gJUcca"
  >
    <input
      class="sc-gsDKAQ lhwlFl"
      data-testid="description-input"
      name="description"
      placeholder="description"
      type="text"
      value=""
    />
    <input
      class="sc-gsDKAQ lhwlFl"
      data-testid="amount-input"
      min="0"
      name="amount"
      placeholder="amount"
      type="number"
      value="0"
    />
    <select
      class="sc-hKwDye husJyr"
      data-testid="currency-input"
      name="currency"
    >
      <option
        value="HUF"
      >
        HUF
      </option>
      <option
        value="USD"
      >
        USD
      </option>
    </select>
    <div
      class="sc-bdvvtL bMEsbY"
      data-testid="save-button"
    >
      Save
    </div>
  </form>
</DocumentFragment>
`;
